iaso/api/user_roles.py

```diff
@@ -0,0 +1,105 @@
+from typing import Any
+from django.shortcuts import get_object_or_404
+from rest_framework.request import Request
+from rest_framework import viewsets, permissions, serializers
+from django.contrib.auth.models import Permission, Group
+from django.db.models import Q, QuerySet
+from rest_framework.response import Response
+from iaso.models import UserRole
+from django.core.paginator import Paginator
+from .common import TimestampField, ModelViewSet
+
+
+class PermissionSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Permission
+        fields = ("id", "name", "codename")
+
+
+class UserRoleSerializer(serializers.ModelSerializer):
+    permissions = serializers.SerializerMethodField("get_permissions")
+    name = serializers.CharField(source="group.name")
+
+    class Meta:
+        model = UserRole
+        fields = ["id", "name", "permissions", "created_at", "updated_at"]
+
+    created_at = TimestampField(read_only=True)
+    updated_at = TimestampField(read_only=True)
+
+    def get_permissions(self, obj):
+        return PermissionSerializer(obj.group.permissions, many=True).data
+
+    def create(self, validated_data):
+        account = self.context["request"].user.iaso_profile.account
+        request = self.context["request"]
+        groupname = request.data.get("name")
+        permissions = request.data.get("permissions", [])
+        if not groupname:
+            return Response({"error": "User group name is required"}, status=400)
+
+        group = Group(name=groupname)
+        group.save()
+
+        if group.id and len(permissions) > 0:
+            for permission_codename in permissions:
+                permission = get_object_or_404(Permission, codename=permission_codename)
+                group.permissions.add(permission)
+            group.save()
+
+        userRole = UserRole.objects.create(group=group, account=account)
+        userRole.save()
+        return userRole
+
+    def update(self, user_role, validated_data):
+        groupname = self.context["request"].data.get("name", None)
+        permissions = self.context["request"].data.get("permissions", None)
+        group = user_role.group
+        if groupname is not None:
+            group.name = groupname
+        if permissions is not None:
+            group.permissions.clear()
+            for permission_codename in permissions:
+                permission = get_object_or_404(Permission, codename=permission_codename)
+                group.permissions.add(permission)
+        group.save()
+        user_role.save()
+        return user_role
+
+
+class HasRolesPermission(permissions.BasePermission):
```

Comment:
See comment below, you can remove this code.

Reply:
Done